code/modules/reagents/chemistry/machinery/chem_dispenser.dm

```diff
@@ -453,17 +460,35 @@
 		parts_rating += servo.tier
 	power_cost = max(new_power_cost, 0.1 KILO WATTS)
 
+/**
+ * Insert, remove, replace the existig beaker. Returns TRUE on success.
+ * Arguments:
+ *
+ * * mob/living/user - the player trying to replace the beaker
+ * * obj/item/reagent_containers/new_beaker - the beaker we are trying to insert, swap with existing or remove if null
+ */
 /obj/machinery/chem_dispenser/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
-	if(!user)
-		return FALSE
-	if(beaker)
+	if(!QDELETED(beaker))
 		try_put_in_hand(beaker, user)
-		beaker = null
-	if(new_beaker)
+	if(!QDELETED(new_beaker))
+		if(!user.transferItemToLoc(new_beaker, src))
+			update_appearance(UPDATE_OVERLAYS)
+			return FALSE
+
 		beaker = new_beaker
-	update_appearance()
+
+	update_appearance(UPDATE_OVERLAYS)
+
 	return TRUE
 
+/// Checks if user can interact with beaker slot and container is valid
+/obj/machinery/chem_dispenser/proc/can_insert_beaker(mob/living/user, obj/item/reagent_containers/container)
+	return is_valid_container(container) && can_interact(user) && user.can_perform_action(src, ALLOW_SILICON_REACH | FORBID_TELEKINESIS_REACH)
+
+/// Checks if container can be used with this machine
+/obj/machinery/chem_dispenser/proc/is_valid_container(obj/item/reagent_containers/container)
+	return is_reagent_container(container) && container.is_open_container() && !(container.item_flags & ABSTRACT) && !(container.flags_1 & HOLOGRAM_1)
```

Comment:
I think these two procs are better off belonging to `obj/item` type than be copypasted on every other chem-related machinery.

On a side note, we should consider splitting the item file at some point, since it's over a thousand lines long now.